### PR TITLE
MTDSA-18949: Parser-validator improvements- RetrieveUkPropertyAnnualSubmission endpoint

### PR DESCRIPTION
**Pull request by KobithasanVasantharajah**

### CODE REVIEW

File: test/v2/controllers/RetrieveUkPropertyAnnualSubmissionControllerSpec.scala [hunk 1]
```diff
@@ -37,49 +37,43 @@ class RetrieveUkPropertyAnnualSubmissionControllerSpec
...
-          .parse(rawData)
-          .returns(Right(requestData))
+        willUseValidator(returningSuccess(requestData))
 
         MockRetrieveUkPropertyService
           .retrieve(requestData)
           .returns(Future.successful(Right(ResponseWrapper(correlationId, responseData))))
 
         MockHateoasFactory
-          .wrap(responseData, hateoasData)
```
Comment (GazalNebHMRC) [inline test/v2/controllers/RetrieveUkPropertyAnnualSubmissionControllerSpec.scala:58]: Can hateoasData be used again here please?

File: test/v2/controllers/RetrieveUkPropertyAnnualSubmissionControllerSpec.scala [hunk 2]
```diff
@@ -37,49 +37,43 @@ class RetrieveUkPropertyAnnualSubmissionControllerSpec
...
+          .wrap(responseData, RetrieveUkPropertyAnnualSubmissionHateoasData(nino, businessId, taxYear))
           .returns(HateoasWrapper(responseData, testHateoasLinks))
 
-        runOkTest(expectedStatus = OK, maybeExpectedResponseBody = Some(responseBodyJsonWithHateoas))
```
Comment (GazalNebHMRC) [inline test/v2/controllers/RetrieveUkPropertyAnnualSubmissionControllerSpec.scala:61]: Better you use responseBodyJsonWithHateoas val here.

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import mocks.MockAppConfig
+import support.UnitSpec
+import v2.models.request.retrieveUkPropertyAnnualSubmission.RetrieveUkPropertyAnnualSubmissionRequestData
+
+class RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec extends UnitSpec with MockAppConfig {
+
+  val validNino: String              = "AA123456B"
+  val validBusinessId: String        = "XAIS12345678901"
+  val validTaxYear: String           = "2021-22"
+  implicit val correlationId: String = "X-123"
```
Comment (GazalNebHMRC) [inline test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:30]: Can correlationId be moved on line 26 to maintain consistency with other endpoints?

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+  private val validatorFactory = new RetrieveUkPropertyAnnualSubmissionValidatorFactory(mockAppConfig)
+
+  private def validator(nino: String, businessId: String, taxYear: String) =
+    validatorFactory.validator(nino, businessId, taxYear)
+
+  class SetUp {
+
+    MockAppConfig.minimumTaxV2Uk
+      .returns(2021)
```
Comment (GazalNebHMRC) [inline test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:44]: minimumTaxV2Uk should return 2022, not 2021 (as per the values in application.conf)

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+      .anyNumberOfTimes()
+
+  }
+
+  "validator" should {
+    "return a request object" when {
```
Comment (GazalNebHMRC) [inline test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:50]: Can this be changed to "return the parsed domain object" instead?

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      "valid request data is supplied" in new SetUp {
+        val result: Either[ErrorWrapper, RetrieveUkPropertyAnnualSubmissionRequestData] =
+          validator(validNino, validBusinessId, validTaxYear).validateAndWrapResult()
+
+        result shouldBe Right(RetrieveUkPropertyAnnualSubmissionRequestData(parsedNino, parsedBusinessId, parsedTaxYear))
+      }
+    }
+
+    "return an ErrorWrapper" when {
```
Comment (GazalNebHMRC) [inline test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:59]: Please change this 'when' test statement on line 59 to "return a single validation error" and put all the trests that return single errors under it. 
Also, please change the test statement on line 60 to be "an invalid nino is supplied".

**Review by GazalNebHMRC — CHANGES_REQUESTED** [review]

File: test/v2/connectors/RetrieveUkPropertyAnnualSubmissionConnectorSpec.scala [hunk 7]
```diff
@@ -31,8 +31,8 @@ import scala.concurrent.Future
 
 class RetrieveUkPropertyAnnualSubmissionConnectorSpec extends ConnectorSpec {
 
-  private val nino: String       = "AA123456A"
-  private val businessId: String = "XAIS12345678910"
+  private val nino: Nino             = Nino("AA123456A")
+  private val businessId: BusinessId = BusinessId("XAIS12345678910")
```
Comment (SamuelJacobsHMRC) [inline test/v2/connectors/RetrieveUkPropertyAnnualSubmissionConnectorSpec.scala:35]:
```suggestion
  private val nino             = Nino("AA123456A")
  private val businessId  = BusinessId("XAIS12345678910")
```

File: test/v2/connectors/RetrieveUkPropertyAnnualSubmissionConnectorSpec.scala [hunk 8]
```diff
@@ -31,8 +31,8 @@ import scala.concurrent.Future
...
 
   private val ukFhlProperty: UkFhlProperty       = UkFhlProperty(None, None)
   private val ukNonFhlProperty: UkNonFhlProperty = UkNonFhlProperty(None, None)
```
Comment (SamuelJacobsHMRC) [inline test/v2/connectors/RetrieveUkPropertyAnnualSubmissionConnectorSpec.scala:38]:
```suggestion
  private val ukFhlProperty        = UkFhlProperty(None, None)
  private val ukNonFhlProperty = UkNonFhlProperty(None, None)
```

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala (hunk 3 above)
Comment (SamuelJacobsHMRC) [inline test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:30]: can make all these vals private and can remove the type of `String`

File: test/v2/services/RetrieveUkPropertyAnnualSubmissionServiceSpec.scala [hunk 9]
```diff
@@ -17,24 +17,24 @@
...
 class RetrieveUkPropertyAnnualSubmissionServiceSpec extends UnitSpec {
 
-  private val nino: String       = "AA123456A"
-  private val businessId: String = "XAIS12345678910"
-  private val taxYear: TaxYear   = TaxYear.fromMtd("2020-21")
+  private val nino: Nino             = Nino("AA123456A")
+  private val businessId: BusinessId = BusinessId("XAIS12345678910")
+  private val taxYear: TaxYear       = TaxYear.fromMtd("2020-21")
 
   implicit private val correlationId: String = "X-123"
```
Comment (SamuelJacobsHMRC) [inline test/v2/services/RetrieveUkPropertyAnnualSubmissionServiceSpec.scala:39]:
```suggestion
  private val nino             = Nino("AA123456A")
  private val businessId  = BusinessId("XAIS12345678910")
  private val taxYear        = TaxYear.fromMtd("2020-21")

  implicit private val correlationId = "X-123"
```

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+      val result: Either[ErrorWrapper, RetrieveUkPropertyAnnualSubmissionRequestData] =
+        validator(validNino, validBusinessId, "2021-23").validateAndWrapResult()
+
+      result shouldBe Left(ErrorWrapper(correlationId, RuleTaxYearRangeInvalidError))
+    }
+
+    "return multiple validation errors" when {
+      "the request has multiple issues (path parameters)" in new SetUp {
+        val result: Either[ErrorWrapper, RetrieveUkPropertyAnnualSubmissionRequestData] =
+          validator("Nino", "BusinessId", "2021/22").validateAndWrapResult()
```
Comment (SamuelJacobsHMRC) [inline test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:101]: for the sake of consistency with the other tests in the other PRs and for explicitness can we do:
```suggestion
          validator("invalid", "invalid", "invalid").validateAndWrapResult()
```

**Review by SamuelJacobsHMRC — CHANGES_REQUESTED** [review]

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala (hunk 4 above)
Comment (KobithasanVasantharajah) [reply test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:44]: This should be 2021- take a look at test/v2/controllers/requestParsers/validators/RetrieveUkPropertyAnnualSubmissionValidatorSpec.scala which is what was used to create this ValidatorFactorySpec.

File: test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala (hunk 3 above)
Comment (SamuelJacobsHMRC) [reply test/v2/controllers/validators/RetrieveUkPropertyAnnualSubmissionValidatorFactorySpec.scala:30]: didn't mean to include the correlationId, that needs to remain typed as a String as its an implicit

**Review by GazalNebHMRC — DISMISSED** [review]

**Review by GazalNebHMRC — APPROVED** [review]

**Review by SamuelJacobsHMRC — APPROVED** [review]

**Pull request merged**